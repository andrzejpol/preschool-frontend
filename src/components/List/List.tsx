import React from "react";

const List = ({columns, renderRow, data}: {
  columns: { header: string; accessor: string; className?: string }[];
  renderRow: (item: any) => React.ReactNode;
  data: any[]
}) => {

  const noData = data?.length == 0;

  return (
    <table className="w-full mt-4">
      <thead>
      <tr className="text-left text-gray-500 text-sm">
        {columns.map(col => (
          <th key={col.accessor} className={col.className}>{col.header}</th>
        ))}
      </tr>
      </thead>
      {noData ? (<tbody className="flex items-center justify-center h-64 w-full ">
      <tr>
        <td colSpan={4}><p className="text-xl font-semibold text-gray-500">There is no data</p></td>
      </tr>
      </tbody>) : (<tbody>{data.map((item) => renderRow(item))}</tbody>)}
    </table>
  )
};

export default List;