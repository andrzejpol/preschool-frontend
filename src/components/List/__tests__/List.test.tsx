import React from "react";
import {render, screen} from "@testing-library/react";
import List from "../List";

describe("List Component", () => {
  const columns = [
    {header: "Name", accessor: "name", className: "text-left"},
    {header: "Age", accessor: "age", className: "text-center"},
    {header: "Email", accessor: "email", className: "text-right"},
  ];

  const renderRow = (item) => (
    <tr key={item.id}>
      <td>{item.name}</td>
      <td>{item.age}</td>
      <td>{item.email}</td>
    </tr>
  );

  const data = [
    {id: 1, name: "John Doe", age: 30, email: "john@example.com"},
    {id: 2, name: "Jane Smith", age: 25, email: "jane@example.com"},
  ];

  it("renders the table headers correctly", () => {
    render(<List columns={columns} renderRow={renderRow} data={data}/>);

    columns.forEach((col) => {
      expect(screen.getByText(col.header)).toBeInTheDocument();
    });
  });

  it("displays no data message when data is empty", () => {
    render(<List columns={columns} renderRow={renderRow} data={[]}/>);

    expect(screen.getByText("There is no data")).toBeInTheDocument();
  });

  it("renders the table rows correctly when data is provided", () => {
    render(<List columns={columns} renderRow={renderRow} data={data}/>);

    data.forEach((item) => {
      expect(screen.getByText(item.name)).toBeInTheDocument();
      expect(screen.getByText(item.age.toString())).toBeInTheDocument();
      expect(screen.getByText(item.email)).toBeInTheDocument();
    });
  });

  it("matches the snapshot when data is present", () => {
    const {asFragment} = render(
      <List columns={columns} renderRow={renderRow} data={data}/>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("matches the snapshot when no data is present", () => {
    const {asFragment} = render(
      <List columns={columns} renderRow={renderRow} data={[]}/>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
