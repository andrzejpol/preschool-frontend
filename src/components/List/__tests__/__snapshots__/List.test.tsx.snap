// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List Component > matches the snapshot when data is present 1`] = `
<DocumentFragment>
  <table
    class="w-full mt-4"
  >
    <thead>
      <tr
        class="text-left text-gray-500 text-sm"
      >
        <th
          class="text-left"
        >
          Name
        </th>
        <th
          class="text-center"
        >
          Age
        </th>
        <th
          class="text-right"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          John Doe
        </td>
        <td>
          30
        </td>
        <td>
          john@example.com
        </td>
      </tr>
      <tr>
        <td>
          Jane Smith
        </td>
        <td>
          25
        </td>
        <td>
          jane@example.com
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`List Component > matches the snapshot when no data is present 1`] = `
<DocumentFragment>
  <table
    class="w-full mt-4"
  >
    <thead>
      <tr
        class="text-left text-gray-500 text-sm"
      >
        <th
          class="text-left"
        >
          Name
        </th>
        <th
          class="text-center"
        >
          Age
        </th>
        <th
          class="text-right"
        >
          Email
        </th>
      </tr>
    </thead>
    <tbody
      class="flex items-center justify-center h-64 w-full "
    >
      <tr>
        <td
          colspan="4"
        >
          <p
            class="text-xl font-semibold text-gray-500"
          >
            There is no data
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
